pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +38,155 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	repo, repoFound := RepositoryFrom(ctx)
+	if !repoFound {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
+	}
+
+	// if there is a repo layer link, return its descriptor
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
+	if err == nil {
+		if repoFound {
+			// and remember the association
+			repo.cachedLayers.RememberDigest(dgst, repo.blobrepottl, imageapi.DockerImageReference{
+				Namespace: repo.namespace,
+				Name:      repo.name,
+			}.Exact())
+		}
+		return desc, err
+	}
+
+	if !repoFound {
+		return distribution.Descriptor{}, distribution.ErrBlobUnknown
+	}
+
+	// verify the blob is stored locally
+	desc, err = dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	if err != nil {
+		return desc, err
+	}
+
+	if imageStreamHasBlob(repo, dgst) {
+		return desc, nil
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+func (bs *blobDescriptorService) Clear(ctx context.Context, dgst digest.Digest) error {
+	if repo, found := RepositoryFrom(ctx); found {
+		repo.cachedLayers.ForgetDigest(dgst, imageapi.DockerImageReference{
+			Namespace: repo.namespace,
+			Name:      repo.name,
+		}.Exact())
+	} else {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
```

Comment:
and if so, should we still be calling `bs.BlobDescriptorService.Clear(ctx, dgst)`? what does clearing the descriptor do? would that affect other users of this blob?